File: src/EligibilityChecker.java
public class EligibilityChecker {
    public boolean canTake (Student student,Unit targetUnit, TeachingPeriod teachingPeriod) {
        boolean prerequisiteSatisfied = targetUnit.getPrerequisite() == null || targetUnit.getPrerequisite().isSatisfied(student);
        boolean corequisiteSatisfied = targetUnit.getCorequisite() == null || targetUnit.getCorequisite().isSatisfied(student);
        boolean prohibitionSatisfied = targetUnit.getProhibition() == null || !targetUnit.getProhibition().isSatisfied(student);
        return  
                prerequisiteSatisfied 
                && corequisiteSatisfied 
                && prohibitionSatisfied 
                && targetUnit.getTeachingPeriod().contains(teachingPeriod);
    }

    public boolean canPlace(
    Student student,
    PlannedUnit plannedUnit
) {
    if (plannedUnit.isForced()) {
        return true;
    }

    return canTake(
        student,
        plannedUnit.getUnit(),
        plannedUnit.getTeachingPeriod()
    );
}
}

File: src/eligibility_checker.cpp
#include "eligibility_checker.hpp"

#include "planned_unit.hpp"
#include "student.hpp"
#include "unit.hpp"

namespace course_planner {

bool EligibilityChecker::can_take(const Student& student,
                                  const Unit& target_unit,
                                  TeachingPeriod teaching_period) const {
    const auto& prerequisite = target_unit.prerequisite();
    const auto& corequisite = target_unit.corequisite();
    const auto& prohibition = target_unit.prohibition();

    const bool prerequisite_satisfied =
        !prerequisite || prerequisite->is_satisfied(student);
    const bool corequisite_satisfied =
        !corequisite || corequisite->is_satisfied(student);
    const bool prohibition_satisfied =
        !prohibition || !prohibition->is_satisfied(student);
    const bool is_offered =
        target_unit.teaching_periods().count(teaching_period) != 0;

    return prerequisite_satisfied && corequisite_satisfied &&
           prohibition_satisfied && is_offered;
}

bool EligibilityChecker::can_place(const Student& student,
                                   const PlannedUnit& planned_unit) const {
    if (planned_unit.is_forced()) {
        return true;
    }

    return can_take(student, planned_unit.unit(),
                    planned_unit.teaching_period());
}

}  // namespace course_planner
